### PR TITLE
set cache to retrieve values up until a day

### PR DESCRIPTION
**Pull request by seunomonije**

### CODE REVIEW

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 1]
```diff
@@ -32,6 +34,8 @@
 /** Implementation of server side cache that stores API requests to save time and API quota. */
 public class AnalysisCache {
   private static final String CACHE_FILE = "cachedData.txt";
+  private static final long ONE_DAY_IN_NANOSECONDS = (long)(8.64 * Math.pow(10, 13));
```
Comment (laptou) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:37]: nit: you can write `8.64e13` to accomplish the same thing

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 2]
```diff
@@ -100,11 +104,23 @@ public static void saveCache() throws InvalidKeyException, IllegalBlockSizeExcep
     }
   }
 
+  public static void cleanDayOldEntries() {
+    long now = Instant.now().getEpochSecond();
+
+    for (Map.Entry cachePair : cacheMap.entrySet()) { 
+      CacheValue currentCachedData = (CacheValue)cachePair.getValue();
+      long timeOfCacheEntry = currentCachedData.timestamp.getEpochSecond();
+      if (now - timeOfCacheEntry >= ONE_DAY_IN_NANOSECONDS) {
```
Comment (laptou) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:113]: `getEpochSecond()` returns time in seconds. Why are you comparing it to `ONE_DAY_IN_NANOSECONDS`?

Comment (seunomonije) [reply server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:113]: Forgot to change one day in nano var..good catch

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 3]
```diff
@@ -100,11 +104,23 @@ public static void saveCache() throws InvalidKeyException, IllegalBlockSizeExcep
     }
   }
 
+  public static void cleanDayOldEntries() {
+    long now = Instant.now().getEpochSecond();
+
+    for (Map.Entry cachePair : cacheMap.entrySet()) { 
+      CacheValue currentCachedData = (CacheValue)cachePair.getValue();
+      long timeOfCacheEntry = currentCachedData.timestamp.getEpochSecond();
+      if (now - timeOfCacheEntry >= ONE_DAY_IN_SECONDS) {
```
Comment (gracetian6) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:113]: nice constant name

File: server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java [hunk 4]
```diff
@@ -66,7 +76,15 @@ protected void doGet(HttpServletRequest req, HttpServletResponse res)
 
     VideoAnalysis servletResults =
         new VideoAnalysis(perspectiveMap, commentsSentiment, commentArray, videoId, videoInfo);
-    AnalysisCache.add(userInput, servletResults);
+    
+    // Only add to the cache if the video is more than 10 days old,
+    // and there are at least 20 comments
+    long now = Instant.now().getEpochSecond();
+    long instantVideoWasPublished = videoInfo.publishedDate.getEpochSecond();
+    if (now - instantVideoWasPublished > 10*ONE_DAY_IN_SECONDS
```
Comment (gracetian6) [inline server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java:84]: nit add space: `10 * ONE_DAY_IN_SECONDS`

Comment (gracetian6) [reply server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java:84]: great comment. If statement makes sense

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 5]
```diff
@@ -32,6 +34,8 @@
 /** Implementation of server side cache that stores API requests to save time and API quota. */
 public class AnalysisCache {
   private static final String CACHE_FILE = "cachedData.txt";
+  private static final long ONE_DAY_IN_SECONDS = 86400;
```
Comment (laptou) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:37]: nit: `ONE_DAY_IN_SECONDS = 24 * 60 * 60`

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 6]
```diff
@@ -100,11 +104,23 @@ public static void saveCache() throws InvalidKeyException, IllegalBlockSizeExcep
     }
   }
 
+  public static void cleanDayOldEntries() {
```
Comment (jessicaslaughter) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:107]: this method needs a comment (same with all public methods)

File: server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java [hunk 7]
```diff
@@ -18,13 +18,14 @@
...
 import java.util.regex.Pattern; 
+import java.time.Instant;
 
 @WebServlet("/api/analysis")
 public class AnalysisServlet extends HttpServlet {
   @Override
   protected void doGet(HttpServletRequest req, HttpServletResponse res)
       throws ServletException, IOException {
-
+    long ONE_DAY_IN_SECONDS = 86400;
```
Comment (jessicaslaughter) [inline server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java:28]: same as Ibiyemi noted above, spread this number out into the equation that resulted in it

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java (hunk 5 above)
Comment (geekster777) [reply server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:37]: Please make this change.

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 8]
```diff
@@ -56,7 +60,11 @@ public static AnalysisCache getInstance()
     }
     return AnalysisCacheObject;
   }
-
+  
+  /**
+   * Loads the cache hashmap from the cache file
+   *
```
Comment (geekster777) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:66]: nit: remove blank line

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 9]
```diff
@@ -84,6 +92,10 @@ public static void loadCache()
     }
   }
 
+  /**
+   * Saves and writes the cache hashmap into the file
+   *
```
Comment (geekster777) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:97]: Same

File: server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java [hunk 10]
```diff
@@ -100,15 +112,48 @@ public static void saveCache() throws InvalidKeyException, IllegalBlockSizeExcep
     }
   }
 
+  /**
+   * Cleans entries that are a day or more old
+   *
```
Comment (geekster777) [inline server/src/main/java/com/google/musicanalysis/cache/AnalysisCache.java:117]: Same

File: server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java [hunk 11]
```diff
@@ -66,7 +76,15 @@ protected void doGet(HttpServletRequest req, HttpServletResponse res)
...
     VideoAnalysis servletResults =
         new VideoAnalysis(perspectiveMap, commentsSentiment, commentArray, videoId, videoInfo);
-    AnalysisCache.add(userInput, servletResults);
+    
+    // Only add to the cache if the video is more than 10 days old,
+    // and there are at least 20 comments
+    long now = Instant.now().getEpochSecond();
+    long instantVideoWasPublished = videoInfo.publishedDate.getEpochSecond();
+    if (now - instantVideoWasPublished > 10*ONE_DAY_IN_SECONDS 
+          && commentArray.size() == 20 ) {
```
Comment (geekster777) [inline server/src/main/java/com/google/musicanalysis/site/AnalysisServlet.java:85]: This 10 and 20 should be stored as file level consts like:
```
MIN_FRESHNESS_TO_CACHE = 10 * ONE_DAY_IN_SECONDS;
MIN_COMMENT_ACTIVITY_TO_CACHE = 20;
```

**Pull request merged**